app/models.py

```diff
@@ -20,14 +22,112 @@
+        The object representing the current user.
+        Can be None only if the use_auto_verify
+        parameter is set to False.
+    use_auto_verify: boolean
+        if True (default), will call verify and commit
+        functions of the session before exiting the context.
+        If True, must specify a legal User object.
+
+    Returns
+    -------
+    a scoped session object that can be used in a context
+    manager to access the database. If auto verify is enabled,
+    will use the current user given to apply verification
+    and commit to the database when exiting context.
+
+    """
+
+    def verify(self):
+        # import here to avoid circular import loop
+        from .handlers.base import bulk_verify
+
+        """Check that the current user has permission to create, read,
+        update, or delete rows that are present in the session. If not,
+        raise an AccessError (causing the transaction to fail and the API to
+        respond with 401).
+
+        """
+        # get items to be inserted
+        new_rows = [row for row in self.new]
+
+        # get items to be updated
+        updated_rows = [row for row in self.dirty if self.is_modified(row)]
+
+        # get items to be deleted
+        deleted_rows = [row for row in self.deleted]
+
+        # get items that were read
+        read_rows = [
+            row
+            for row in set(self.identity_map.values())
+            - (set(updated_rows) | set(new_rows) | set(deleted_rows))
+        ]
+
+        # need to check delete permissions before flushing, as deleted records
+        # are not present in the transaction after flush (thus can't be used in
+        # joins). Read permissions can be checked here or below as they do not
+        # change on flush.
+        for mode, collection in zip(
+            ["read", "update", "delete"],
+            [read_rows, updated_rows, deleted_rows],
+        ):
+            bulk_verify(mode, collection, self.user_or_token)
+
+        # update transaction state in DB, but don't commit yet. this updates
+        # or adds rows in the database and uses their new state in joins,
+        # for permissions checking purposes.
+        self.flush()
+        bulk_verify("create", new_rows, self.user_or_token)
+
+    if use_auto_verify and user_or_token is None:
+        raise RuntimeError(
+            "Cannot start a session with use_auto_verify without a valid user."
+        )
+    if user_or_token is not None and not isinstance(user_or_token, (User, Token)):
+        raise ValueError(
+            "user_or_token must be an instance of User or Token, "
+            f"got {user_or_token.__class__.__name__}."
+        )
+    with scoped_session(sessionmaker(), scopefunc=session_context_id.get)() as session:
+        session.user_or_token = user_or_token
+        session.use_auto_verify = use_auto_verify
+        # make this an instance method of session
+        session.verify = types.MethodType(verify, session)
```

Comment:
Why not just assign the function?

Reply:
Because I want the function to take the `session` as the `self` argument.